Add IPOPT Optimizer

schmoelder opened the pull request:
This PR adds support for interior point optimzer IPOPT using [cyipopt](https://github.com/mechmotum/cyipopt).

@flo-schu I guess this still needs some work but it's been in a local branch for a long time and maybe it would be a good opportunity to include it in the test runners and maybe even for the optimization study.

Could you please jog my memory and briefly tell me what is required to run the `test_optimizer_behaviour` using pytest?

schmoelder committed:
Add method to calculate volumetric flow rate from velocity.

schmoelder committed:
Update docstrings

schmoelder committed:
Add method to create fraction

schmoelder committed:
Add `start` and `end` times to `Fraction`

schmoelder committed:
Specify and validate reference type for difference metrics

schmoelder committed:
Add FractionationSSE

schmoelder committed:
Add FractionationReference

schmoelder committed:
Add classes to __all__

schmoelder committed:
Formatting

schmoelder committed:
Add `only_transforms_solution` flag

schmoelder committed:
Only resample if flag is set

schmoelder committed:
Fix typo

schmoelder committed:
Add IPOPT

Built on cyipopt.Problem, which provides a per-iteration intermediate()
hook with get_current_iterate() for live progress and checkpointing,
matching the other adapters. Gradients and constraint Jacobians are
supplied analytically via objective_jacobian and
nonlinear_constraint_jacobian, keeping per-test runtime in seconds and
ensuring IPOPT receives derivatives in transformed space.

schmoelder committed:
[WIP] Add IPOPT

flo-schu commented:
The new optimizer needs to be inserted here:

```python
@pytest.fixture(params=[
    SLSQP,
    TrustConstr,
    U_NSGA3,
    GPEI,
    NEHVI,
])
def optimizer(request):
    return request.param()
```

that's it. Then call the test with `pytest tests/test_optimizer_behavior.py::test_from_initial_values`

By the way, in `dev` currently both, `test_from_initial_values` and `test_convergence` are used. Both tests are almost the same and I don't think the `test_convergence` test is necessary.

ronald-jaepel committed:
Rename .t0 to .calculate_interstitial_rt in Cstr
Fix and extend tests about .calculate_interstitial_rt/velocity

schmoelder committed:
Unify calling evaluation functions for individuals and populations

Previously, there were two interfaces in the `OptimizationProblem` for calling evaluation functions (e.g. objectives): one for evaluating individuals, and one for populations.
To simplify the code base, these two methods were now unified.
To ensure backward compatibility, a 1D-Array is returned if a single individual is passed to the function.

ronald-jaepel committed:
Add qNParEGO Ax MOO Interface

ronald-jaepel committed:
Relax tolerance for MOO convergence test

ronald-jaepel committed:
Ensure callback dirs are created for final callback

ronald-jaepel committed:
Always inherit cadet path

Previously, the cadet path set in Cadet(install_path="path") was not inherited into cadet instances created from the run() method.

schmoelder committed:
Fix add_concentration_profile (#140)

Co-authored-by: r.jaepel <ronald.jaepel@gmail.com>

schmoelder commented:
@hannahlanzrath, this is another optimizer that uses gradient information. This is WIP, but maybe you'd be up for testing it a bit, especially when considering how Jacobians (and Hesse) matrices computed?

schmoelder commented:
@hannahlanzrath this is the optimizer we discussed with Anton earlier this week. I just realized that somehow you were already assigned. Is this something you would like to explore? I think it's not too complicated and might be a good challenge to better get to know the optimization module in CADET-Process. Obviously, this can wait until after your thesis is handed in.

hannahlanzrath commented:
Yeah, Looks interesting. Maybe something that can also be combined with the inclusion of parameter sensitivities.
